src/Language/DifferentialDatalog/FlatBuffer.hs

```diff
@@ -799,51 +871,282 @@ mkJavaBuilder prog_name =
+
+typeFlatbufJavaReader _ = Nothing
+
+-- Extract value from context.
+-- Java expression 'e' contains a FlatBuffers-native Java type that represents
+-- 'fbctx' context.  Convert it to the corresponding convenience type.
+jReadField :: (?d::DatalogProgram, ?prog_name::String) => Int -> FBCtx -> Doc -> Type -> Doc
+jReadField nesting fbctx e t =
+    case fbctx of
+         -- Anything that's not a table must be wrapped in a table to be stored
+         -- in a union
+         FBUnion     | typeIsTable t
+                     -> do_read e
+                     | otherwise
+                     -> jReadField (nesting+1) (FBField (typeTableName t) "v") e t
+         -- Anything that's not a union or array can be stored in an array; unions and arrays must be
+         -- wrapped in a table first.
+         FBArray     | typeHasUniqueConstructor t && not (typeIsVector t)
+                     -> do_read e
+                     | otherwise
+                     -> jReadField (nesting+1) (FBField (typeTableName t) "v") e t
+         -- Vectors are representedFlatBuffers has a special method for embedding vectors in a buffer
+         FBField{..} | typeIsVector t
+                     -> do_read (e <> "." <> pp ( fbctxField))
+                     | not (typeHasUniqueConstructor t)
+                     -> do_read (e <> "." <> jAccessorName fbctxField <> "Type()," <+>
+                                 "(Table t) ->" <+> e <> "." <> jAccessorName fbctxField <> "(t)")
+                     | otherwise
+                     -> do_read (e <> "." <> jAccessorName fbctxField <> "()")
+    where
+    do_read e' = case typeNormalizeForFlatBuf t of
+                      TBool{}            -> e'
+                      TInt{}             -> bigint
+                      TString{}          -> e'
+                      TBit{..} | typeWidth <= 64
+                                         -> (parens $ jConvTypeR t) <> e'
+                      TBit{..}           -> biguint
+                      TSigned{..} | typeWidth <= 64
+                                         -> (parens $ jConvTypeR t) <> e'
+                      TSigned{..}        -> bigint
+                      TTuple{..}         -> "new" <+> jConvTypeR t <> parens e'
+                      TUser{..} |typeHasUniqueConstructor t
+                                         -> "new" <+> jConvTypeR t <> parens e'
+                                | otherwise
+                                         -> jConvTypeR t <> ".init" <> parens e'
+                      TOpaque{typeArgs = [elemType], ..} | elem typeName sET_TYPES ->
+                          let elem_type = jConvObjTypeR elemType
+                              ltype = "java.util.List<" <> elem_type <> ">"
+                              -- Add nesting depth to local variables inside lambda so that they don't
+                              -- clash with variables from containing scopes.
+                              vlen = "len" <> pp nesting
```

Comment:
this is a new scope, it shouldn't be a problem I would think - unless you want to refer to outside variables.